### PR TITLE
Ddcnls 7087

### PR DESCRIPTION
**Pull request by brendanstraughan**

### CODE REVIEW

File: app/controllers/BulkRequestReceivedController.scala [hunk 1]
```diff
@@ -39,24 +39,22 @@ class BulkRequestReceivedController @Inject()(authAction: AuthAction,
...
                   case OK => Ok(views.html.bulk_request_received(bulkRequest.reference))
                   case CONFLICT => Ok(views.html.failure(Messages("gmp.bulk.failure.duplicate_upload"), Messages("gmp.bulk.problem.header"), Messages("gmp.bulk_failure_duplicate.title")))
                   case REQUEST_ENTITY_TOO_LARGE => Ok(views.html.failure(Messages("gmp.bulk.failure.too_large"), Messages("gmp.bulk.file_too_large.header"), Messages("gmp.bulk_failure_file_too_large.title")))
                   case _ => Ok(views.html.failure(Messages("gmp.bulk.failure.generic"), Messages("gmp.bulk.problem.header"), Messages("gmp.bulk_failure_generic.title")))
                 }
 
-              case Right(e: DataLimitExceededException) =>
-                Future.successful(Ok(views.html.failure(Messages("gmp.bulk.failure.too_large"), Messages("gmp.bulk.file_too_large.header"), Messages("gmp.bulk_failure_file_too_large.title"))))
+              case Right(e: DataLimitExceededException) => Future.successful(Ok(views.html.failure(Messages("gmp.bulk.failure.too_large"), Messages("gmp.bulk.file_too_large.header"), Messages("gmp.bulk_failure_file_too_large.title"))))
```
Comment (oc-ste) [inline app/controllers/BulkRequestReceivedController.scala]: duplicated code with the Left case: REQUEST_ENTITY_TOO_LARGE

File: app/controllers/BulkRequestReceivedController.scala [hunk 2]
```diff
@@ -39,24 +39,22 @@ class BulkRequestReceivedController @Inject()(authAction: AuthAction,
...
+
+              case Right(_) => Future.successful((Redirect(controllers.routes.IncorrectlyEncodedController.get())))
```
Comment (oc-ste) [inline app/controllers/BulkRequestReceivedController.scala]: Missing a test for this case

File: app/controllers/BulkRequestReceivedController.scala [hunk 3]
```diff
@@ -39,24 +39,22 @@ class BulkRequestReceivedController @Inject()(authAction: AuthAction,
   def get = authAction.async {
       implicit request => {
         val link = request.linkId
-
         Logger.debug(s"[BulkRequestReceivedController][get][GET] : ${request.body}")
         sessionService.fetchGmpBulkSession().flatMap {
           case Some(session) if session.callBackData.isDefined =>
             val callbackData = session.callBackData.get
-
             bulkRequestCreationService.createBulkRequest(callbackData.collection, callbackData.id, session.emailAddress.getOrElse(""), session.reference.getOrElse("")) match {
```
Comment (oc-ste) [inline app/controllers/BulkRequestReceivedController.scala]: Consider refactoring createBulkRequest to work with standard paradigm where Left contains the error cases and Right is the happy path.

File: app/controllers/BulkRequestReceivedController.scala [hunk 4]
```diff
@@ -39,24 +39,22 @@ class BulkRequestReceivedController @Inject()(authAction: AuthAction,
...
 
-              case Left(bulkRequest) =>
-                gmpBulkConnector.sendBulkRequest(bulkRequest, link).map {
+              case Left(bulkRequest) => gmpBulkConnector.sendBulkRequest(bulkRequest, link).map {
                   case OK => Ok(views.html.bulk_request_received(bulkRequest.reference))
                   case CONFLICT => Ok(views.html.failure(Messages("gmp.bulk.failure.duplicate_upload"), Messages("gmp.bulk.problem.header"), Messages("gmp.bulk_failure_duplicate.title")))
                   case REQUEST_ENTITY_TOO_LARGE => Ok(views.html.failure(Messages("gmp.bulk.failure.too_large"), Messages("gmp.bulk.file_too_large.header"), Messages("gmp.bulk_failure_file_too_large.title")))
                   case _ => Ok(views.html.failure(Messages("gmp.bulk.failure.generic"), Messages("gmp.bulk.problem.header"), Messages("gmp.bulk_failure_generic.title")))
```
Comment (oc-ste) [inline app/controllers/BulkRequestReceivedController.scala]: Should this be the same as the Right(_) case? It's confusing that it renders a generic failure for Left but Right has a redirect

File: conf/messages [hunk 5]
```diff
@@ -14,6 +14,12 @@ gmp.bulk_request_received.title = GMP bulk file received - GOV.UK
 gmp.bulk.results.title = Download GMP calculation - GOV.UK
 gmp.bulk.results_not_found.title = GMP calculation no longer available - GOV.UK
 gmp.bulk.wrong_user.title = Wrong GMP checker login - GOV.UK
+
+gmp.bulk.incorrectlyEncoded = Your file is incorrectly encoded. To get a calculation it must be UTF-8 encoded (modern standard).
+gmp.bulk.incorrectlyEncoded.header = Sorry, there is a problem
+gmp.bulk_failure_incorrectlyEncoded.title = Sorry, there is a problem
+
```
Comment (oc-ste) [inline conf/messages]: These are the same, you can juse use one

File: app/views/incorrectlyEncoded.scala.html [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+@import uk.gov.hmrc.play.views.html._
+@import uk.gov.hmrc.play.views.html.helpers._
+@import views.html.helpers._
+
+@gmp_main(title = header + " - " + Messages("service.title") + " - " + Messages("gov.uk"), supportLinkEnabled = false) {
+
+@includes.back_link(header)
+<h1>@header</h1>
+<br/>
```
Comment (oc-ste) [inline app/views/incorrectlyEncoded.scala.html]: think br tags are not allowed in our service, best check with Mo

File: app/services/BulkRequestCreationService.scala [hunk 7]
```diff
@@ -89,26 +90,31 @@ class BulkRequestCreationService @Inject()( environment: Environment,
...
+          }
+          else {
+            val req = BulkCalculationRequest(id, email, reference, enterLineNumbers(bulkCalculationRequestLines.drop(1)))
+            Logger.debug(s"[BulkRequestCreationService][createBulkRequest] size : ${req.calculationRequests.size}")
+            Left(req)
+          }
+        }
+
+
+      case Failure(_) => Right(new IncorrectlyEncodedException)
```
Comment (oc-ste) [inline app/services/BulkRequestCreationService.scala]: Missing a test for this case.
Consider logging the exception returned here

File: app/views/incorrectlyEncoded.scala.html (hunk 6 above)
Comment (brendanstraughan) [reply app/views/incorrectlyEncoded.scala.html]: done

File: app/services/BulkRequestCreationService.scala (hunk 7 above)
Comment (brendanstraughan) [reply app/services/BulkRequestCreationService.scala]: done

**Pull request merged**